### PR TITLE
Publish user events

### LINKED ISSUE #6325: Event Publishing: User Events
Support publishing of user activity events:
- user login
- user logout
- user changed password
- admin changes user's password

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes #6325 
# What's Changed

# Screenshots

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/users/forms.py [hunk 1]
```diff
@@ -70,3 +73,12 @@ def clean(self):
...
 class PreferenceProfileSettingsForm(BootstrapMixin, forms.Form):
     timezone = TimeZoneFormField(required=False, help_text="Set your default timezone", widget=StaticSelect2)
+
+
+class AdminPasswordChangeForm(_AdminPasswordChangeForm):
+    def save(self, commit):
+        # Override `_AdminPasswordChangeForm.save()` to publish admin change user password event
+        instance = super().save(commit)
+        payload = {"data": serialize_object_v2(instance)}
+        publish_event(topic="nautobot.admin.user.change_password", payload=payload)
```
Comment (glennmatthews) [inline nautobot/users/forms.py:83]: Should this be gated by a `if commit:` check?

File: nautobot/docs/user-guide/platform-functionality/events.md [hunk 2]
```diff
@@ -56,3 +56,15 @@ The data payload associated with events of any of the above topics has the follo
...
 * `postchange` - a dictionary of record attributes and their values *after* the event occurred (or null, in the case of a `nautobot.delete` event)
 * `differences` - a dictionary with keys `added` and `removed`, each of which is a dictionary of record attributes that changed in the event, providing a convenient alternative to manually comparing the prechange and postchange data.
+
+
+### User Events
+
+* `nautobot.users.user.login` - a new record was created
+* `nautobot.users.user.logout` - a new record was created
+* `nautobot.users.user.change_password` - a new record was created
+* `nautobot.admin.user.change_password` - a new record was created
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/events.md:66]: Please update descriptions for accuracy

File: nautobot/docs/user-guide/platform-functionality/events.md [hunk 3]
```diff
@@ -56,3 +56,15 @@ The data payload associated with events of any of the above topics has the follo
...
+
+The data payload associated with events of any of the above topics has the following keys:
+
+ * `data` - A dictionary of the affected user record attributes and their value.
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/events.md:70]: Please provide full example data - in particular, want to be very sure that password information is *not* exposed in the payload

**Review by glennmatthews — COMMENTED** [review]
Please provide example payloads.

File: nautobot/core/tests/test_views.py [hunk 4]
```diff
@@ -399,6 +402,48 @@ def test_api_docs_403_unauthenticated(self):
...
+                    "username": self.user.username,
+                    "password": "pass",
+                },
+            )
+        self.user.refresh_from_db()
+        payload = {"data": serialize_object_v2(self.user)}
+        self.assertEqual(
+            cm.output,
+            [f"INFO:nautobot.events.nautobot.users.user.login:{json.dumps(payload, indent=4)}"],
+        )
```
Comment (glennmatthews) [inline nautobot/core/tests/test_views.py:433]: Maybe add an `assertNotIn("password", cm.output)` and/or `assertNotIn("pass", cm.output)` as well?

File: nautobot/docs/user-guide/platform-functionality/events.md [hunk 5]
```diff
@@ -56,3 +56,40 @@ The data payload associated with events of any of the above topics has the follo
...
+                "username": "admin",
+                "first_name": "",
+                "last_name": "",
+                "email": "a@a.com",
+                "is_staff": true,
+                "is_active": true,
+                "date_joined": "2024-10-22T04:35:28.886682Z",
+                "config_data": {},
+                "groups": [],
+                "default_saved_views": []
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/events.md:91]: Might be nice to exclude `config_data` and `default_saved_views` from the payload, even though they're a default part of the serializer they're irrelevant to the event.

File: nautobot/users/forms.py [hunk 6]
```diff
@@ -70,3 +73,13 @@ def clean(self):
...
     timezone = TimeZoneFormField(required=False, help_text="Set your default timezone", widget=StaticSelect2)
+
+
+class AdminPasswordChangeForm(_AdminPasswordChangeForm):
+    def save(self, commit, *args, **kwargs):
+        # Override `_AdminPasswordChangeForm.save()` to publish admin change user password event
+        instance = super().save(commit, *args, **kwargs)
+        if commit:
+            payload = {"data": serialize_object_v2(instance)}
+            publish_event(topic="nautobot.admin.user.change_password", payload=payload)
```
Comment (glennmatthews) [inline nautobot/users/forms.py:84]: Any luck with a unit test for this one?

File: nautobot/users/tests/test_views.py [hunk 7]
```diff
@@ -40,6 +43,37 @@ def test_change_password_enabled(self):
...
+                    "old_password": "foo",
+                    "new_password1": "bar",
+                    "new_password2": "bar",
+                },
+            )
+        payload = {"data": serialize_object_v2(self.user)}
+        self.assertEqual(
+            cm.output,
+            [f"INFO:nautobot.events.nautobot.users.user.change_password:{json.dumps(payload, indent=4)}"],
+        )
```
Comment (glennmatthews) [inline nautobot/users/tests/test_views.py:73]: Similarly would be good to assert that neither `password` nor `foo` nor `bar` appear in the event payload.

**Review by glennmatthews — COMMENTED** [review]
@lampwins Do we also need events for any of the following:

- User logs in or out via SSO as opposed to local auth?
- User authenticates to the API via a token?

File: nautobot/users/forms.py (hunk 6 above)
Comment (timizuoebideri1) [reply nautobot/users/forms.py:87]: ive tested it maually, but been having issues writing unittest for it

File: nautobot/core/models/utils.py [hunk 8]
```diff
@@ -150,6 +150,13 @@ def serialize_object_v2(obj):
     return data
 
 
+def serialize_user_without_config_and_views(user):
```
Comment (glennmatthews) [inline nautobot/core/models/utils.py:153]: Any particular reason this is in `core` rather than `users` app?

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**